### ISSUE TITLE
WPF on net6.0 - AsyncRelayCommand blocks UI

### ISSUE THREAD

**Issue by thomasrea0113** [CommunityToolkit/dotnet#218]
### Describe the bug

IAsyncRelayCommand does not run asyncronously. the `IsRunning` property is only updated *after* the task has completed. The task completes, `IsRunning` is set to `true`, and then immediately set to `false`. I verified this by adding a handler to `ProcessCommand.OnPropertyChanged` and can see that the async action completes before `IsRunning` is ever set to true.

All other bindings work as expected. The "process" button enables/disabled based on whether or not a file is selected. 

### Reproducible in sample app?

- [ ] This bug can be reproduced in the sample app.

### Steps to reproduce

```xaml
<Window x:Class="ImportUtility.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:vms="clr-namespace:ImportUtility.ViewModels"
        mc:Ignorable="d"
        Title="Import Utility" Height="450" Width="800" d:DataContext="{d:DesignInstance Type=vms:MainWindowViewModel}">
    <Grid>
        <Button Content="Process Template" Command="{Binding ProcessCommand}" CommandParameter="{Binding TemplateFile}" HorizontalAlignment="Center" Margin="0,164,0,0" VerticalAlignment="Top" Height="96" Width="174"/>
        <Button Content="close" Command="{Binding CancelCommand}" HorizontalAlignment="Center" Margin="0,326,0,0" VerticalAlignment="Top"/>
        <Button Content="Select File" Command="{Binding SelectFile}" HorizontalAlignment="Left" Margin="385,102,0,0" VerticalAlignment="Top"/>
        <Label Content="{Binding TemplateFile}" HorizontalAlignment="Center" Margin="0,141,0,0" VerticalAlignment="Top" Width="174"/>
    </Grid>
</Window>
```

With MainWindow.xaml.cs:

```csharp
    public partial class MainWindow : Window
    {
        public MainWindowViewModel ViewModel => (MainWindowViewModel)DataContext;

        public MainWindow()
        {
            InitializeComponent();
            DataContext = Ioc.Default.GetRequiredService<MainWindowViewModel>();
        }
    }
```

And ViewModel:

```csharp

public class MainWindowViewModel : ViewModelBase
{
    private readonly IIUTemplateProcessor _templateProcessor;
    private readonly string _uploadFolder;

    private string? _templateFile;
    public string? TemplateFile
    {
        get => _templateFile;
        set => SetProperty(ref _templateFile, value, boundCommands: new[] { ProcessCommand });
    }

    public ICommand SelectFile { get; }

    public IRelayCommand CancelCommand { get; }
    public IAsyncRelayCommand ProcessCommand { get; }

    public MainWindowViewModel(IIUTemplateProcessor templateProcessor,
        [NotNull] IOptionsSnapshot<AppSettings> settings)
    {
        _templateProcessor = templateProcessor;
        _uploadFolder = settings.Value.Folders.Upload.EnsureAppFolder();

        SelectFile = new RelayCommand(() =>
        {
            var dlg = new OpenFileDialog();
            dlg.InitialDirectory = _uploadFolder;
            TemplateFile = dlg.ShowDialog(App.Current.MainWindow) == true ? dlg.FileName : null;
        });

        ProcessCommand = new AsyncRelayCommand<string>(async (file, token) =>
        {
            if (file == null) throw new NullReferenceException(nameof(file));

            try
            {
                var template = new IUTemplateInfo(file);
                await _templateProcessor.ProcessTemplateAsync(template, token: token).ConfigureAwait(true);
            }
            catch (BatchException) { }
        }, f => !string.IsNullOrWhiteSpace(f));

        CancelCommand = new RelayCommand(() => ProcessCommand.Cancel(), () => ProcessCommand.IsRunning);

        ProcessCommand.PropertyChanged += (_, args) =>
        {
            if (args.PropertyName == nameof(ProcessCommand.IsRunning))
                CancelCommand.NotifyCanExecuteChanged();
        };
    }
}
```

Custom ViewModelBase for good measure:

```csharp

public abstract class ViewModelBase : ObservableObject
{
    /// <summary>
    /// Checks if a property already matches a desired value.  Sets the property and
    /// notifies listeners only when necessary.
    /// </summary>
    /// <typeparam name="T">Type of the property.</typeparam>
    /// <param name="storage">Reference to a property with both getter and setter.</param>
    /// <param name="value">Desired value for the property.</param>
    /// <param name="propertyName">Name of the property used to notify listeners.  This
    /// <param name="boundMembers">Additional properties to notify of this change</param>
    /// value is optional and can be provided automatically when invoked from compilers that
    /// support CallerMemberName.</param>
    /// <returns>True if the value was changed, false if the existing value matched the
    /// desired value.</returns>
    protected virtual bool SetProperty<T>(ref T storage, T value,
        string[]? boundMembers = null,
        IRelayCommand[]? boundCommands = null,
        [CallerMemberName] string? propertyName = null)
    {
        var changed = base.SetProperty(ref storage, value, propertyName);

        if (changed)
        {
            if (boundMembers != null)
                foreach (var member in boundMembers)
                    OnPropertyChanged(member);
            if (boundCommands != null)
                foreach (var command in boundCommands)
                    command.NotifyCanExecuteChanged();
        }

        return changed;
    }
}

```

App.xaml:

```xaml
<Application x:Class="ImportUtility.App"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
</Application>
```

### Expected behavior

`IsRunning` should be set to true prior to the async action starting, and the `AsyncRelayCommand` should not block the UI.

### Screenshots

![iasynccommand-blocking](https://user-images.githubusercontent.com/14018447/163581109-42e4833d-7145-420f-be2d-225a31c1860e.gif)

### Windows Build Number

- [ ] Windows 10 1809 (Build 17763)
- [ ] Windows 10 1903 (Build 18362)
- [ ] Windows 10 1909 (Build 18363)
- [ ] Windows 10 2004 (Build 19041)
- [ ] Windows 10 20H2 (Build 19042)
- [ ] Windows 10 21H1 (Build 19043)
- [ ] Windows 11 21H2 (Build 22000)
- [X] Other (specify)

### Other Windows Build number

21H2 19044.1586

### App minimum and target SDK version

- [ ] Windows 10, version 1809 (Build 17763)
- [ ] Windows 10, version 1903 (Build 18362)
- [ ] Windows 10, version 1909 (Build 18363)
- [ ] Windows 10, version 2004 (Build 19041)
- [ ] Other (specify)

### Visual Studio Version

2022

### Help us help you

No.

**Comment by michael-hawker** [comment]
FYI @Sergio0694 another issue from WCT transferred over.

**Comment by thomasrea0113** [comment]
```csproj
<PackageReference Include="Microsoft.Toolkit.Mvvm" Version="7.1.2" />
```

**Comment by mikechristiansenvae** [comment]
> but it seems that the IU will block until await is actually called within the delegate. So in the above example, the IU will unblock at the call to ProcessTemplateAsync.

That is how `async` and `await` work.

[From the documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/async) (emphasis mine):

> **An async method runs synchronously until it reaches its first await expression**, at which point the method is suspended until the awaited task is complete. In the meantime, control returns to the caller of the method, as the example in the next section shows.

> However in my scenario, Instantiating the IUTemplateInfo class is quite slow, and causes a several second delay on very large templates.

I suggest that you use the factory pattern.  Create an interface (and/or an appropriate concrete class)


```
public interface IUITemplateInfoFactory
{
    public Task<IUTemplateInfo> CreateTemplateAsync(IFileInfo file);
}
```

Then your code becomes:

```
        ProcessCommand = new AsyncRelayCommand<string>((file, token) => Task.Run(async () =>
        {
            if (file == null) throw new NullReferenceException(nameof(file));

            try
            {
                var template = await factory.CreateTemplateAsync(file);
                await _templateProcessor.ProcessTemplateAsync(template, token: token).ConfigureAwait(false);
            }
            catch (BatchException) { }
        }, token), f => !string.IsNullOrWhiteSpace(f) && ProcessCommand?.IsRunning == false);
```

And now your method will only be synchronous until its finished with its null check - it'll be async after that.

**Comment by thomasrea0113** [comment]
> > but it seems that the IU will block until await is actually called within the delegate. So in the above example, the IU will unblock at the call to ProcessTemplateAsync.
> 
> That is how `async` and `await` work.
> 
> [From the documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/async) (emphasis mine):
> 
> > **An async method runs synchronously until it reaches its first await expression**, at which point the method is suspended until the awaited task is complete. In the meantime, control returns to the caller of the method, as the example in the next section shows.
> 
> > However in my scenario, Instantiating the IUTemplateInfo class is quite slow, and causes a several second delay on very large templates.
> 
> I suggest that you use the factory pattern. Create an interface (and/or an appropriate concrete class)
> 
> ```
> public interface IUITemplateInfoFactory
> {
>     public Task<IUTemplateInfo> CreateTemplateAsync(IFileInfo file);
> }
> ```
> 
> Then your code becomes:
> 
> ```
>         ProcessCommand = new AsyncRelayCommand<string>((file, token) => Task.Run(async () =>
>         {
>             if (file == null) throw new NullReferenceException(nameof(file));
> 
>             try
>             {
>                 var template = await factory.CreateTemplateAsync(file);
>                 await _templateProcessor.ProcessTemplateAsync(template, token: token).ConfigureAwait(false);
>             }
>             catch (BatchException) { }
>         }, token), f => !string.IsNullOrWhiteSpace(f) && ProcessCommand?.IsRunning == false);
> ```
> 
> And now your method will only be synchronous until its finished with its null check - it'll be async after that.

Thank you, I appreciate the clarification. I figured it was a misunderstanding on my part. I'll get this issue closed.

**Comment by mikechristiansenvae** [comment]
@thomasrea0113 - I know it's been quite a while, but I saw this issue again, and I remembered something....

Instead of using the factory pattern, like I recommended ☝️, you could simply use `await Task.Yield();`


```csharp
ProcessCommand = new AsyncRelayCommand<string>((file, token) => Task.Run(async () =>
        {
            if (file == null) throw new NullReferenceException(nameof(file));
            await Task.Yield();  // <------------------- Add this line right here
            try
            {
                var template = new IUTemplateInfo(file);
                await _templateProcessor.ProcessTemplateAsync(template, token: token).ConfigureAwait(false);
            }
            catch (BatchException) { }
        }, token), f => !string.IsNullOrWhiteSpace(f) && ProcessCommand?.IsRunning == false);
```

As I mentioned above, the method is synchronous until the first `await` is encountered.  By using `await Task.Yield();`, you give it an `await` at the top of the method.

`Task.Yield()` doesn't _do_ anything, in and of itself - it just yields control back, and then "waits in line" for async operations.
